### ISSUE TITLE
dynamic adaption to webgl context deconstruction

### ISSUE THREAD

**Issue by dkoes** [3dmol/3Dmol.js#292]
I would like it if when a 3Dmol viewer's webgl context is disabled (usually due to limits) the view was automatically replaced with an image.  Then, if the image is clicked on, the context gets reinstantiated with the model data.

I'm not actually sure this is possible.

**Comment by sjaustirni** [comment]
Are you talking about [Handling Context Lost](https://www.khronos.org/webgl/wiki/HandlingContextLost)?

And what image are you talking about? Just a general placeholder image, or the last saved frame?

**Comment by dkoes** [comment]
Yes, although the main reason we lose a context is browsers have a hard limit on the number of simultaneous contexts.  The grid viewer helps ameliorate this problem (provides multiple viewers in one context), but sometimes people want many distinct viewers on the page.

The image would be a picture of the context before it is disabled.

**Comment by sjaustirni** [comment]
It seems like it should be theoretically possible ([demo for one context in a page that preserves the buffer when the context is lost](https://github.com/davidwparker/programmingtil-webgl/blob/91dd9ca3b720f35b2c89953e71fe5e8f9a6b8d5f/0093-lost-context-part-2/index.js#L295)), although due to the hard browsers limit this may not necessarily work in this case and more investigation is required.

**Comment by dkoes** [comment]
You either need to get an event from the browser that the context is about to be destroyed (which I haven't been able to find) or you have to poll, which is undesirable for performance reasons.

**Comment by sjaustirni** [comment]
Is the drawing buffer from previous complete frame invalidated right upon the context loss?

I was thinking that if the drawing buffers are not invalidated immediately the above linked method may work, because it just reuses the last successfully rendered frame. And if many contexts are needed, the last frame can always be converted to an image and then displayed.

I was not able to find the answer to this vital question though and don't have time for experimenting today. I may investigate this tomorrow though, if this functionality is still needed.

**Comment by sjaustirni** [comment]
OK, I did not do much experimentation, but I did some more reading.

According to [WebGL spec](https://www.khronos.org/registry/webgl/specs/latest/1.0/), the previous drawing buffer is discarded on `webglcontextrestored` event, see **5.15.3 The Context Restored Event, no. 3** (my emphasis):
>  Create a drawing buffer using the settings specified in context's context creation parameters, and associate the drawing buffer with context, _discarding any previous drawing buffer_. 

Also checkout that the demo whose [code](https://github.com/davidwparker/programmingtil-webgl/blob/91dd9ca3b720f35b2c89953e71fe5e8f9a6b8d5f/0093-lost-context-part-2/index.js#L295) I have previously linked [actually works](https://youtu.be/wGoJN2Dci9E?t=131).

**Comment by dkoes** [comment]
This page demonstrates the issue:
http://3dmol.csb.pitt.edu/tests/multi.html

The earliest contexts are destroyed and replaced with a frowny face (on chrome).

**Comment by dkoes** [comment]
Robustness to arbitrary loss of contexts would be good too though.

**Comment by sjaustirni** [comment]
I have realized that it is not exactly possible to "trigger" context restore. It is an event that is fired by the browser at its will. Whilst we could make it such that the context restore does nothing and wait until the user click to start re-instantiation, this does not really solve the problem with too many canvases, as in that case context restore is never fired.

I suggest to stop caring about context restore event altogether and have one way of handling lost context. 

I have made a pull request #399 which simply appends new canvas in the place of old (and removes the old) upon user click. The drawing frame seems to be preserved in the newest Firefox, but Chrome does no such thing.

I admit this only solves half of the issue, but maybe it's good enough (?). I would love if you tried my PR out and told me what you think :D
